rxandroid/src/main/java/rx/android/schedulers/HandlerScheduler.java

```diff
@@ -44,45 +45,49 @@ public Worker createWorker() {
     }
 
     static class HandlerWorker extends Worker {
-
-        final Handler handler;
-
-        private final CompositeSubscription compositeSubscription = new CompositeSubscription();
+        private final Handler handler;
+        /* Writes are guarded by 'this'. */
+        private volatile boolean unsubscribed;
 
         HandlerWorker(Handler handler) {
             this.handler = handler;
         }
 
         @Override
         public void unsubscribe() {
-            compositeSubscription.unsubscribe();
+            synchronized (this) {
+                if (!unsubscribed) {
+                    unsubscribed = true;
+                    handler.removeCallbacksAndMessages(this /* token */);
+                }
+            }
         }
 
         @Override
         public boolean isUnsubscribed() {
-            return compositeSubscription.isUnsubscribed();
+            return unsubscribed;
         }
 
         @Override
         public Subscription schedule(Action0 action, long delayTime, TimeUnit unit) {
-            if (compositeSubscription.isUnsubscribed()) {
+            if (unsubscribed) {
                 return Subscriptions.unsubscribed();
             }
 
             action = RxAndroidPlugins.getInstance().getSchedulersHook().onSchedule(action);
 
-            final ScheduledAction scheduledAction = new ScheduledAction(action);
-            scheduledAction.addParent(compositeSubscription);
-            compositeSubscription.add(scheduledAction);
+            ScheduledAction scheduledAction = new ScheduledAction(action, handler);
 
-            handler.postDelayed(scheduledAction, unit.toMillis(delayTime));
+            Message message = Message.obtain(handler, scheduledAction);
+            message.obj = this; // Used as token for unsubscription operation.
 
-            scheduledAction.add(Subscriptions.create(new Action0() {
-                @Override
-                public void call() {
-                    handler.removeCallbacks(scheduledAction);
+            synchronized (this) {
```

Comment:
Does this have to be synchronized? Why not do the post-check after the submission and remove token+message pair?